lib/elixir/lib/map_set.ex

```diff
@@ -63,6 +63,30 @@ defmodule MapSet do
   # with Elixir 1.14 and earlier and executed on Elixir 1.15+.
   defstruct map: :sets.new(version: 2)
 
+  @doc """
+  Checks if `map_set` contains `value`. Allowed in guards.
+
+  Otherwise equivalent to `member?/2`.
+
+  ## Examples
+
+      iex> MapSet.is_member(MapSet.new([1, 2, 3]), 2)
+      true
+
+      iex> MapSet.is_member(MapSet.new([1, 2, 3]), 4)
+      false
+
+      iex> case MapSet.new([1, 2, 3]) do
+      ...>   map_set when MapSet.is_member(map_set, 3) -> true
+      ...>   _ -> false
+      ...> end
+      true
+
```

Comment:
Should we include:

```suggestion

Returns `false` if `map_set` is not a `MapSet`:

      iex> MapSet.is_member([1, 2, 3], 2)
      false

```

Reply:
Hrm, good point, but it should rather raise:

```
iex(1)> is_map_key([], 123)
** (BadMapError) expected a map, got: []
    :erlang.is_map_key(123, [])
    iex:1: (file)
```

Usually we would do `is_struct(...) or :fail`. But I think the error my be unreadable. Now I am pretty sure we discussed this in the past and this was one of the blockers? Or maybe we implement it using a macro instead of `defguard` and we check the context accordingly.

Reply:
Macro sounds good?

Reply:
Wasn't a part of those discussions, so I'm not clear on the desired failure modes. Do we want:

- If `map_set` is not a map
  - If in a guard
    - `is_struct(...) or :fail`, to short-circuit out of the guard
  - If outside a guard
    - raise a new `BadMapSetError` type thing

Reply:
aka something like https://github.com/elixir-lang/elixir/commit/3443473cec07690db403272582b191deaa96a90f

Reply:
> raise a new `BadMapSetError` type thing

Perhaps just an `ArgumentError` is fine in this case, since it is not a primitive type and has no need for [normalizing erlang errors](https://github.com/elixir-lang/elixir/blob/main/lib/elixir/lib/exception.ex#L2343-L2349)? I'm not sure.

Reply:
Yeah, I think you're right—`ArgumentError` is the idiomatic exception to use here.

Worth pointing out that (`:sets.new(version:2)`)'s `:sets.is_element/2` and `MapSet.member?` both throw `FunctionClauseError` in this case. Not something we can readily raise here, but worth noting the difference in documentation.

Reply:
Actually, probably not worth documenting—that way we're not committed to the exact error raised in either `is_member/2` or `member?/2` and can always normalize later.

Reply:
Resolved in [8459b05](https://github.com/elixir-lang/elixir/pull/13386/commits/8459b05c5d4cebc894d9482bf2e026a2c57b9db9)

Reply:
Ah, [this PR](https://github.com/elixir-lang/elixir/pull/7063) makes the case for not raising in guards, I see what the original concerns here are. Will change and add to tests.

Reply:
Resolved in 61bc1a2edf9f13c5f591f9514d0af1c8c277201c

Reply:
This actually makes me question if `Integer.is_{even|odd}/2` should be implemented similarly, to raise an `ArgumentError` on wrong type input.

Reply:
This is a good question, I'm wondering what is preferable:
- being consistent between guards and regular function call
- being assertive and fail

knowing that assertive guards could only fail silently...
Consistency might be nice because you can try out the behavior in IEx and know it is going to be how your guard behaves (I think at least for `is_even` we should keep it as is?)